docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+
+- `unknown` is the universal supertype i.e. the widest possible type, equivalent to the universal set(U).
+- Every type is assignable to `unknown`, but `unknown` is only assignable to `unknown`.
+- When typing the _assignee_, `any` and `unknown` are completely interchangeable since every type is assignable to both.
+
+###### `never`
+
+`never` is worth trying as a starting point for narrowing down the type, as it is the universal subtype and assignable to all types.
+
+- `never` is the universal subtype i.e. the narrowest possible type, equivalent to the null set(∅).
+- `never` is assignable to every type, but the only type that is assignable to `never` is `never`.
+- When typing the _assigned_:
+  - `unknown` is unable to replace `any`, as `unknown` is only assignable to `unknown`.
+  - The type of the _assigned_ must be a subtype of the _assignee_.
+  - If the assigned type simultaneously needs to be the assignee for another type, `never` will not work, as no type is assignable to `never`.
+
+<!-- TODO: Add examples -->
+
+##### Don't allow `any` to be used as a generic default argument
+
+Some generic types use `any` as a default generic argument. This can silently introduce an `any` type into the code, causing unexpected behavior and suppressing useful errors.
+
+🚫
+
+```typescript
+const NETWORKS = new Map({
+  mainnet: '0x1',
+  goerli: '0x5',
+}); // Type 'Map<any, any>'
+
+const mockGetNetworkConfigurationByNetworkClientId = jest.fn(); // Type 'jest.Mock<any, any>'
+mockGetNetworkConfigurationByNetworkClientId.mockImplementation(
+  (origin, type) => {},
+); // No error!
+// Even though 'mockImplementation' should only accept callbacks with a signature of '(networkClientId: string) => NetworkConfiguration | undefined'
+```
+
+✅
+
+```typescript
+const NETWORKS = new Map<string, `0x${string}`>({
+  mainnet: '0x1',
+  goerli: '0x5',
+}); // Type 'Map<string, `0x${string}`>'
+
+const mockGetNetworkConfigurationByNetworkClientId = jest.fn<
+  ReturnType<NetworkController['getNetworkConfigurationByNetworkClientId']>,
+  Parameters<NetworkController['getNetworkConfigurationByNetworkClientId']>
+>(); // Type 'jest.Mock<NetworkConfiguration | undefined, [networkClientId: string]>'
+mockGetNetworkConfigurationByNetworkClientId.mockImplementation(
+  (origin, type) => {},
+);
+// Argument of type '(origin: any, type: any) => void' is not assignable to parameter of type '(networkClientId: string) => NetworkConfiguration | undefined'.
+// Target signature provides too few arguments. Expected 2 or more, but got 1.ts(2345)
+```
+
+#### Acceptable usages of `any`
+
+`any` is only acceptable for very specific usages. The following should be considered an exhaustive list of its valid use cases.
+
+##### `any` may be necessary when assigning new properties to or deleting properties from a generic type at runtime
```

Comment:
Hmm, I don't think I've encountered this situation before. Has this come up? I'm not sure why we'd want to do this.

Reply:
This is related to a specific error I encountered on the `any` ticket for rate-limit-controller, and the example is [here](https://github.com/MetaMask/contributor-docs/pull/74/files/e6114410c2115c6783e0ccece94cbbaa336df6db#diff-ae03a9f2eb638a25c1d8d7ba853c78220f1ed44a19ad97a8d780cbc0c00a4260R528-R541). I'm realizing the order of presentation might be a little confusing on these examples hmm...

Reply:
Oh I see, sorry no I did see the example but for some reason I assumed it was an "artificial" situation written to demonstrate this point. But I see that it is a real example.

I thought this guideline was about adding or deleting properties in a way that is disallowed by the types. But I guess it's meant for situations where the type allows it, but TypeScript doesn't properly "collapse" the generic to the specific example the code applies to.

Reply:
I'm curious about what the limitation is here, do you know if it's the same as what this bug describes? https://github.com/microsoft/TypeScript/issues/36631
Is there another workaround that we could consider here?

But more generally, do you think we should encourage `@ts-expect-error` here or a type assertion, rather than `any`? At least it would be less "infectious".

Reply:
I think this is a different issue. In that bug, the index signature is not being recognized as valid, which is why the workaround works by constraining the type of `K`. 

In this example, the index signature is valid, but the issue is it being used in a write operation on a generic type.

On the subject of what our recommendation should be...

`as any` doesn't seem to be "infectious" like `any`. It only applies to the target variable and not to other instances of the same variable. I should add a sentence about this, but without necessarily encouraging its use.

`@ts-expect-error` is probably the best option here. There might be an argument for conceptually distinguishing `@ts-expect-error` and `any` use cases. But from a pragmatic viewpoint, I think preferring `@ts-expect-error` in all cases makes sense because of how it alerts us when the error it's suppressing is no longer there.

I'll move this over to the `@ts-expect-error` section if that sounds ok.

Reply:
Sounds good! That makes sense to me

Reply:
I ended up removing the "assignment to generics" example and creating a new "Prefer `as unknown as` over `as any`" entry: https://github.com/MetaMask/contributor-docs/pull/74/commits/aadb1051362540afe53be54acb6d45d718dc2470.

I'm not sure if the generics assignment is the best example to present as a `@ts-expect-error` use case. I'll bring it back if I can find a good slot for it.